src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/Streaming/DataStreamWriterTests.cs

```diff
@@ -59,5 +63,190 @@ public void TestSignaturesV2_3_X()
+            // is rethrown by ForeachWriterWrapper. We will limit the partitions
+            // to 1 to make validating this scenario simpler. Verify that:
+            // - Triggering stream throws an exception.
+            // - 1 CSV file will be created in the temporary directory.
+            // - 1 Exception will be created in the temporary directory. The
+            //   thrown exception from Process() will be sent to Close().
+            // - The CSV file will not contain valid data to read.
+            TestAndValidateForeach(
+                streamInputPath,
+                new TestForeachWriterProcessFailure(),
+                1,
+                1,
+                Enumerable.Empty<int>());
+        }
+
+        private void TestAndValidateForeach(
+            string streamInputPath,
+            TestForeachWriter foreachWriter,
+            int expectedCSVFiles,
+            int expectedExceptionFiles,
+            IEnumerable<int> expectedOutput)
+        {
+            // Temporary folder the TestForeachWriter will write to.
+            using var dstTempDirectory = new TemporaryDirectory();
+            foreachWriter.WritePath = dstTempDirectory.Path;
+
+            // Read streamInputPath, repartition data, then
+            // call TestForeachWriter on the data.
+            DataStreamWriter dsw = _spark
+                .ReadStream()
+                .Schema("id INT")
+                .Json(streamInputPath)
+                .Repartition(expectedCSVFiles)
+                .WriteStream()
+                .Foreach(foreachWriter);
+
+            // Trigger the stream batch once.
+            if (expectedExceptionFiles > 0)
+            {
+                Assert.Throws<Exception>(
+                    () => dsw.Trigger(Trigger.Once()).Start().AwaitTermination());
+            }
+            else
+            {
+                dsw.Trigger(Trigger.Once()).Start().AwaitTermination();
+            }
+
+            // Verify that TestForeachWriter created a unique .csv when
+            // ForeachWriter.Open was called on each partitionId.
+            Assert.Equal(
+                expectedCSVFiles,
+                Directory.GetFiles(dstTempDirectory.Path, "*.csv").Length);
+
+            // Only if ForeachWriter.Process(Row) throws an exception, will
+            // ForeachWriter.Close(Exception) create a file with the
+            // .exeception extension.
+            Assert.Equal(
+                expectedExceptionFiles,
+                Directory.GetFiles(dstTempDirectory.Path, "*.exception").Length);
+
+            // Read in the *.csv file(s) generated by the TestForeachWriter.
+            // If there are multiple input files, sorting by "id" will make
+            // validation simpler. Contents of the *.csv will only be populated
+            // on successful calls to the ForeachWriter.Process method.
+            DataFrame foreachWriterOutputDF = _spark
+                .Read()
+                .Schema("id INT")
```

Comment:
can we have a test with more complex type? (multi columns, at least one col is struct type, etc.)

Reply:
Added more columns and also a `StructType` as a column.